src/Microsoft.ML.Data/Training/LearnerInputBase.cs

```diff
@@ -0,0 +1,197 @@
+        internal CachingOptions Caching = CachingOptions.Auto;
+    }
+
+    /// <summary>
+    /// The base class for all learner inputs that support a Label column.
+    /// </summary>
+    [TlcModule.EntryPointKind(typeof(CommonInputs.ITrainerInputWithLabel))]
+    public abstract class LearnerInputBaseWithLabel : LearnerInputBase
+    {
+        /// <summary>
+        /// Column to use for labels.
+        /// </summary>
+        [Argument(ArgumentType.AtMostOnce, HelpText = "Column to use for labels", ShortName = "lab", SortOrder = 3, Visibility = ArgumentAttribute.VisibilityType.EntryPointsOnly)]
+        public string LabelColumn = DefaultColumnNames.Label;
+    }
+
+    // REVIEW: This is a known antipattern, but the solution involves the decorator pattern which can't be used in this case.
+    /// <summary>
+    /// The base class for all learner inputs that support a weight column.
+    /// </summary>
+    [TlcModule.EntryPointKind(typeof(CommonInputs.ITrainerInputWithWeight))]
+    public abstract class LearnerInputBaseWithWeight : LearnerInputBaseWithLabel
+    {
+        /// <summary>
+        /// Column to use for example weight.
+        /// </summary>
+        [Argument(ArgumentType.AtMostOnce, HelpText = "Column to use for example weight", ShortName = "weight", SortOrder = 4, Visibility = ArgumentAttribute.VisibilityType.EntryPointsOnly)]
+        public string WeightColumn = null;
+    }
+
+    /// <summary>
+    /// The base class for all unsupervised learner inputs that support a weight column.
+    /// </summary>
+    [TlcModule.EntryPointKind(typeof(CommonInputs.IUnsupervisedTrainerWithWeight))]
+    public abstract class UnsupervisedLearnerInputBaseWithWeight : LearnerInputBase
+    {
+        /// <summary>
+        /// Column to use for example weight.
+        /// </summary>
+        [Argument(ArgumentType.AtMostOnce, HelpText = "Column to use for example weight", ShortName = "weight", SortOrder = 4, Visibility = ArgumentAttribute.VisibilityType.EntryPointsOnly)]
+        public string WeightColumn = null;
+    }
+
+    [TlcModule.EntryPointKind(typeof(CommonInputs.ITrainerInputWithGroupId))]
+    public abstract class LearnerInputBaseWithGroupId : LearnerInputBaseWithWeight
+    {
+        /// <summary>
+        /// Column to use for example groupId.
+        /// </summary>
+        [Argument(ArgumentType.AtMostOnce, Name = "GroupIdColumn", HelpText = "Column to use for example groupId", ShortName = "groupId", SortOrder = 5, Visibility = ArgumentAttribute.VisibilityType.EntryPointsOnly)]
+        public string GroupIdColumn = null;
+    }
+
+    [BestFriend]
+    internal static class LearnerEntryPointsUtils
+    {
+        public static string FindColumn(IExceptionContext ectx, DataViewSchema schema, Optional<string> value)
```

Comment:
>FindColumn [](start = 29, length = 10)

why you moving this one?

Reply:
it depends on LearnerInput* classes

Reply:
it depends on LearnerInput* classes

---
In reply to: [260525594](https://github.com/dotnet/machinelearning/pull/2748#discussion_r260525594) [](ancestors = 260525594)